eth.c

```diff
@@ -0,0 +1,210 @@
+		struct connected_eth_core *core_info = &wh_dev->connected_eth_cores[wh_dev->num_connected_cores];
+		u32 x = WH_ETH_NOC0_X[eth_channel];
+		u32 y = WH_ETH_NOC0_Y[eth_channel];
+		u32 fw_version = wh_noc_read32(tt, x, y, ETH_FW_VERSION_ADDR);
+		u32 port_status = wh_noc_read32(tt, x, y, ETH_PORT_STATUS_ADDR + (eth_channel * 4));
+		u32 remote_rack = wh_noc_read32(tt, x, y, ETH_REMOTE_RACK_ADDR);
+		u32 remote_shelf = wh_noc_read32(tt, x, y, ETH_REMOTE_SHELF_ADDR);
+		u32 rack_shelf = wh_noc_read32(tt, x, y, ETH_LOCAL_RACK_SHELF_ADDR);
+
+		if (fw_version < ETH_MIN_FW_VERSION)
+			continue;
+
+		if (port_status == ETH_STATUS_UNKNOWN || port_status == ETH_STATUS_NOT_CONNECTED)
+			continue;
+
+		core_info->core_num = eth_channel;
+		core_info->fw_version = fw_version;
+
+		core_info->local.rack_x = (rack_shelf >> 0) & 0xFF;
+		core_info->local.rack_y = (rack_shelf >> 8) & 0xFF;
+		core_info->local.shelf_x = (rack_shelf >> 16) & 0xFF;
+		core_info->local.shelf_y = (rack_shelf >> 24) & 0xFF;
+		core_info->local_noc_x = x;
+		core_info->local_noc_y = y;
+
+		core_info->remote.rack_x = (remote_rack >> 0) & 0xFF;
+		core_info->remote.rack_y = (remote_rack >> 8) & 0xFF;
+		core_info->remote.shelf_x = (remote_shelf >> 16) & 0x3F;
+		core_info->remote.shelf_y = (remote_shelf >> 22) & 0x3F;
+		core_info->remote_noc_x = (remote_shelf >> 4) & 0x3F;
+		core_info->remote_noc_y = (remote_shelf >> 10) & 0x3F;
+
+		wh_dev->num_connected_cores++;
+	}
+}
+
+bool wormhole_remote_read32(struct wormhole_device *wh_dev, u32 eth_channel, struct eth_addr_t *eth_addr, u32 noc_x,
+			    u32 noc_y, u64 addr, u32 *value)
+{
+	struct tenstorrent_device *tt = &wh_dev->tt;
+	struct eth_cmd_t cmd;
+	unsigned long timeout;
+	u32 x = WH_ETH_NOC0_X[eth_channel];
+	u32 y = WH_ETH_NOC0_Y[eth_channel];
+	u32 req_rd, req_wr, resp_rd, resp_wr; // queue pointers; 0 <= ptr < 8
+	u32 req_slot, resp_slot; // queue indices; 0 <= slot < 4
+	u32 req_offset, resp_offset;
+	u32 resp_flags_offset, resp_data_offset;
+	u32 fw_version;
+	u32 resp_flags = 0;
+
+	fw_version = wh_noc_read32(tt, x, y, ETH_FW_VERSION_ADDR);
+	if (fw_version < ETH_MIN_FW_VERSION) {
+		pr_err("ETH FW version: %u is too old.\n", fw_version);
+		goto wormhole_eth_read32_done;
+	}
+
+	// Read the current position of the read and write pointers for both the
+	// request and response queues.
+	req_wr = wh_noc_read32(tt, x, y, ETH_REQ_WR_PTR_ADDR);
+	req_rd = wh_noc_read32(tt, x, y, ETH_REQ_RD_PTR_ADDR);
+	resp_wr = wh_noc_read32(tt, x, y, ETH_RESP_WR_PTR_ADDR);
+	resp_rd = wh_noc_read32(tt, x, y, ETH_RESP_RD_PTR_ADDR);
+
+	// Encode the command.
+	memset(&cmd, 0, sizeof(struct eth_cmd_t));
+	cmd.sys_addr = encode_sys_addr(eth_addr->shelf_x, eth_addr->shelf_y, noc_x, noc_y, addr);
+	cmd.data = sizeof(u32);
+	cmd.rack = ((u16)eth_addr->rack_y) << 8 | eth_addr->rack_x;
+	cmd.flags = ETH_CMD_RD_REQ;
+
+	if (eth_queue_full(req_wr, req_rd)) {
+		pr_err("ETH queue %u full\n", eth_channel);
```

Comment:
I think this can get very spammy? You might want to look into the ratelimit helpers.